File: claude-code-agents-python/src/models/language.py
"""Language configuration models for multi-language codegen."""

from __future__ import annotations

from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field

__all__ = [
    "SupportedLanguage",
    "LSP_BINARIES",
    "SDK_LANGUAGES",
    "LanguageConfig",
]


class SupportedLanguage(str, Enum):
    """All supported programming languages for code generation."""

    PYTHON = "python"
    TYPESCRIPT = "typescript"
    GO = "go"
    RUST = "rust"
    JAVA = "java"
    KOTLIN = "kotlin"
    SWIFT = "swift"
    CSHARP = "csharp"
    PHP = "php"
    RUBY = "ruby"
    ELIXIR = "elixir"
    SCALA = "scala"


LSP_BINARIES: dict[SupportedLanguage, str] = {
    SupportedLanguage.PYTHON: "pyright",
    SupportedLanguage.TYPESCRIPT: "typescript-language-server",
    SupportedLanguage.GO: "gopls",
    SupportedLanguage.RUST: "rust-analyzer",
    SupportedLanguage.JAVA: "jdtls",
    SupportedLanguage.KOTLIN: "kotlin-language-server",
    SupportedLanguage.SWIFT: "sourcekit-lsp",
    SupportedLanguage.CSHARP: "OmniSharp",
    SupportedLanguage.PHP: "phpactor",
    SupportedLanguage.RUBY: "solargraph",
    SupportedLanguage.ELIXIR: "elixir-ls",
    SupportedLanguage.SCALA: "metals",
}


SDK_LANGUAGES: dict[SupportedLanguage, str] = {
    SupportedLanguage.PYTHON: "anthropic",
    SupportedLanguage.TYPESCRIPT: "@anthropic-ai/sdk",
    SupportedLanguage.GO: "github.com/anthropics/anthropic-sdk-go",
    SupportedLanguage.RUST: "anthropic (crate)",
    SupportedLanguage.JAVA: "com.anthropic:anthropic-java",
    SupportedLanguage.KOTLIN: "com.anthropic:anthropic-java",
    SupportedLanguage.SWIFT: "anthropic-swift",
    SupportedLanguage.CSHARP: "Anthropic.SDK",
    SupportedLanguage.PHP: "anthropic-php",
    SupportedLanguage.RUBY: "anthropic-rb",
    SupportedLanguage.ELIXIR: "anthropic_ex",
    SupportedLanguage.SCALA: "com.anthropic:anthropic-java",
}


class LanguageConfig(BaseModel):
    """Configuration for a specific language in code generation."""

    language: SupportedLanguage = Field(..., description="The programming language")
    lsp_binary: str = Field(default="", description="LSP server binary name")
    sdk_package: str = Field(default="", description="Anthropic SDK package identifier")
    file_extensions: list[str] = Field(
        default_factory=list, description="Common file extensions"
    )
    build_tool: str = Field(default="", description="Primary build tool")
    test_command: Optional[str] = Field(
        default=None, description="Default test command"
    )
    format_command: Optional[str] = Field(
        default=None, description="Default format command"
    )

    @classmethod
    def for_language(cls, lang: SupportedLanguage) -> LanguageConfig:
        """Create a LanguageConfig with defaults for the given language."""
        configs: dict[SupportedLanguage, dict] = {
            SupportedLanguage.PYTHON: {
                "file_extensions": [".py", ".pyi"],
                "build_tool": "pip",
                "test_command": "pytest",
                "format_command": "ruff format .",
            },
            SupportedLanguage.TYPESCRIPT: {
                "file_extensions": [".ts", ".tsx", ".mts"],
                "build_tool": "npm",
                "test_command": "npm test",
                "format_command": "npx prettier --write .",
            },
            SupportedLanguage.GO: {
                "file_extensions": [".go"],
                "build_tool": "go",
                "test_command": "go test ./...",
                "format_command": "gofmt -w .",
            },
            SupportedLanguage.RUST: {
                "file_extensions": [".rs"],
                "build_tool": "cargo",
                "test_command": "cargo test",
                "format_command": "cargo fmt",
            },
            SupportedLanguage.JAVA: {
                "file_extensions": [".java"],
                "build_tool": "gradle",
                "test_command": "gradle test",
                "format_command": "google-java-format -i **/*.java",
            },
            SupportedLanguage.KOTLIN: {
                "file_extensions": [".kt", ".kts"],
                "build_tool": "gradle",
                "test_command": "gradle test",
                "format_command": "ktlint --format",
            },
            SupportedLanguage.SWIFT: {
                "file_extensions": [".swift"],
                "build_tool": "swift",
                "test_command": "swift test",
                "format_command": "swift-format format -i -r .",
            },
            SupportedLanguage.CSHARP: {
                "file_extensions": [".cs", ".csx"],
                "build_tool": "dotnet",
                "test_command": "dotnet test",
                "format_command": "dotnet format",
            },
            SupportedLanguage.PHP: {
                "file_extensions": [".php"],
                "build_tool": "composer",
                "test_command": "vendor/bin/phpunit",
                "format_command": "vendor/bin/php-cs-fixer fix",
            },
            SupportedLanguage.RUBY: {
                "file_extensions": [".rb", ".rake"],
                "build_tool": "bundler",
                "test_command": "bundle exec rspec",
                "format_command": "bundle exec rubocop -A",
            },
            SupportedLanguage.ELIXIR: {
                "file_extensions": [".ex", ".exs"],
                "build_tool": "mix",
                "test_command": "mix test",
                "format_command": "mix format",
            },
            SupportedLanguage.SCALA: {
                "file_extensions": [".scala", ".sc"],
                "build_tool": "sbt",
                "test_command": "sbt test",
                "format_command": "sbt scalafmtAll",
            },
        }
        defaults = configs.get(lang, {})
        return cls(
            language=lang,
            lsp_binary=LSP_BINARIES.get(lang, ""),
            sdk_package=SDK_LANGUAGES.get(lang, ""),
            file_extensions=defaults.get("file_extensions", []),
            build_tool=defaults.get("build_tool", ""),
            test_command=defaults.get("test_command"),
            format_command=defaults.get("format_command"),
        )

File: claude-code-agents-typescript/src/models/language.ts
// src/models/language.ts — Language configuration for multi-language codegen
import type { LanguageId } from '../types.js';
import { toLanguageId } from '../types.js';

// ── Supported Languages (Discriminated Union via const) ─────────
export const SUPPORTED_LANGUAGES = [
  'python',
  'typescript',
  'go',
  'rust',
  'java',
  'kotlin',
  'swift',
  'csharp',
  'php',
  'ruby',
  'elixir',
  'scala',
] as const;

export type SupportedLanguage = (typeof SUPPORTED_LANGUAGES)[number];

export function isSupportedLanguage(value: string): value is SupportedLanguage {
  return (SUPPORTED_LANGUAGES as readonly string[]).includes(value);
}

// ── LSP Binary Mapping ──────────────────────────────────────────
export const LSP_BINARIES: Readonly<Record<SupportedLanguage, string>> = {
  python: 'pyright',
  typescript: 'typescript-language-server',
  go: 'gopls',
  rust: 'rust-analyzer',
  java: 'jdtls',
  kotlin: 'kotlin-language-server',
  swift: 'sourcekit-lsp',
  csharp: 'OmniSharp',
  php: 'phpactor',
  ruby: 'solargraph',
  elixir: 'elixir-ls',
  scala: 'metals',
};

// ── SDK Package Mapping ─────────────────────────────────────────
export const SDK_PACKAGES: Readonly<Record<SupportedLanguage, string>> = {
  python: 'anthropic',
  typescript: '@anthropic-ai/sdk',
  go: 'github.com/anthropics/anthropic-sdk-go',
  rust: 'anthropic (crate)',
  java: 'com.anthropic:anthropic-java',
  kotlin: 'com.anthropic:anthropic-java',
  swift: 'anthropic-swift',
  csharp: 'Anthropic.SDK',
  php: 'anthropic-php',
  ruby: 'anthropic-rb',
  elixir: 'anthropic_ex',
  scala: 'com.anthropic:anthropic-java',
};

// ── Language Configuration ──────────────────────────────────────
export interface LanguageConfig {
  readonly language: SupportedLanguage;
  readonly languageId: LanguageId;
  readonly lspBinary: string;
  readonly sdkPackage: string;
  readonly fileExtensions: readonly string[];
  readonly buildTool: string;
  readonly testCommand: string | undefined;
  readonly formatCommand: string | undefined;
}

const LANGUAGE_DEFAULTS: Readonly<
  Record<
    SupportedLanguage,
    {
      fileExtensions: readonly string[];
      buildTool: string;
      testCommand: string;
      formatCommand: string;
    }
  >
> = {
  python: {
    fileExtensions: ['.py', '.pyi'],
    buildTool: 'pip',
    testCommand: 'pytest',
    formatCommand: 'ruff format .',
  },
  typescript: {
    fileExtensions: ['.ts', '.tsx', '.mts'],
    buildTool: 'npm',
    testCommand: 'npm test',
    formatCommand: 'npx prettier --write .',
  },
  go: {
    fileExtensions: ['.go'],
    buildTool: 'go',
    testCommand: 'go test ./...',
    formatCommand: 'gofmt -w .',
  },
  rust: {
    fileExtensions: ['.rs'],
    buildTool: 'cargo',
    testCommand: 'cargo test',
    formatCommand: 'cargo fmt',
  },
  java: {
    fileExtensions: ['.java'],
    buildTool: 'gradle',
    testCommand: 'gradle test',
    formatCommand: 'google-java-format -i **/*.java',
  },
  kotlin: {
    fileExtensions: ['.kt', '.kts'],
    buildTool: 'gradle',
    testCommand: 'gradle test',
    formatCommand: 'ktlint --format',
  },
  swift: {
    fileExtensions: ['.swift'],
    buildTool: 'swift',
    testCommand: 'swift test',
    formatCommand: 'swift-format format -i -r .',
  },
  csharp: {
    fileExtensions: ['.cs', '.csx'],
    buildTool: 'dotnet',
    testCommand: 'dotnet test',
    formatCommand: 'dotnet format',
  },
  php: {
    fileExtensions: ['.php'],
    buildTool: 'composer',
    testCommand: 'vendor/bin/phpunit',
    formatCommand: 'vendor/bin/php-cs-fixer fix',
  },
  ruby: {
    fileExtensions: ['.rb', '.rake'],
    buildTool: 'bundler',
    testCommand: 'bundle exec rspec',
    formatCommand: 'bundle exec rubocop -A',
  },
  elixir: {
    fileExtensions: ['.ex', '.exs'],
    buildTool: 'mix',
    testCommand: 'mix test',
    formatCommand: 'mix format',
  },
  scala: {
    fileExtensions: ['.scala', '.sc'],
    buildTool: 'sbt',
    testCommand: 'sbt test',
    formatCommand: 'sbt scalafmtAll',
  },
};

export function languageConfigFor(lang: SupportedLanguage): LanguageConfig {
  const defaults = LANGUAGE_DEFAULTS[lang];
  return {
    language: lang,
    languageId: toLanguageId(lang),
    lspBinary: LSP_BINARIES[lang],
    sdkPackage: SDK_PACKAGES[lang],
    fileExtensions: defaults.fileExtensions,
    buildTool: defaults.buildTool,
    testCommand: defaults.testCommand,
    formatCommand: defaults.formatCommand,
  };
}
